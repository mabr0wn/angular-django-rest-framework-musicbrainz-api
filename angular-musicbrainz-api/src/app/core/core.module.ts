import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BoldPipe } from './pipe/bold.pipe';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';



@NgModule({
  declarations: [BoldPipe],
  imports: [
    CommonModule,
    HttpClientModule
  ],
  exports:[BoldPipe]
})
export class CoreModule { }
