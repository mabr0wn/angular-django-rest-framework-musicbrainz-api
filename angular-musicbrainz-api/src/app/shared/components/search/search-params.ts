export class SearchParams {
    term: string;
    type: string;
    constructor(term, type) {
      this.term = term;
      this.type = type;
    }

    equals(params: SearchParams): boolean {
      return this.term === params.term && this.type === params.type;
    }
  }
