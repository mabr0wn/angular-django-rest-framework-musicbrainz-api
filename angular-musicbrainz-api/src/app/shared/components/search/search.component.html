<input 
  matInput 
  id="search-input" 
  [matAutocomplete]="auto" 
  [formControl]="searchControl" 
  [placeholder]="searchType === 'release' ? 'Album Title' : 'Artist Name'" 
  (keyup)="searchFor(search.value)" 
  name="search" 
  minlength="3" 
  required 
  #search>
<mat-autocomplete  #auto="matAutocomplete">
  <mat-option *ngFor="let item of filteredResults$ | async" [value]="item">
      <span [innerHTML]="item | bold:searchControl.value"></span>
  </mat-option>
</mat-autocomplete>
