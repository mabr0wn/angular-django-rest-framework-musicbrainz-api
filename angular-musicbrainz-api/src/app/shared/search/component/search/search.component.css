@import '@angular/material/prebuilt-themes/deeppurple-amber.css';


:host {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100%;
  } 
  
  h1 {
    font-weight: 300;
    font-size: 40px;
    color: rgba(0, 0, 0, 0.87);
  }
  
  input {
    border: solid 1px rgba(0, 0, 0, 0.1);
    border-radius: 2px;
    outline: none;
    padding: 6px 16px;
  
    width: 100%;
    max-width: 375px;
  
    font-size: 16px;
    line-height: 34px;
  }
  .fa-search {
    margin-left: 27rem;
    margin-top: -2.25rem;
  }
  
  .spacer {
    height: 200px;
  }
  
  