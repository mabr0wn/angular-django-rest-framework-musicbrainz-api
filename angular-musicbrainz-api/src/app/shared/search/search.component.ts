// Angular
import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
// RxJs
import { Observable } from 'rxjs';
import { map, startWith } from 'rxjs/operators';
// Dummy data
import { SAMPLE_RESULTS } from '../../sample-results';


@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css']
})
export class SearchComponent implements OnInit {
  searchControl: FormControl;

  filteredResults$: Observable<string[]>;

  results = SAMPLE_RESULTS;

  constructor() { }

  ngOnInit() {
    this.searchControl = new FormControl('');
    this.filteredResults$ = this.searchControl.valueChanges.pipe(
      startWith(''),
      map(val => this.filterResults(val)),
      map(val => val.slice(0, 4)));
  }

  filterResults(val: string): string[] {
    return val ? this.results.filter(v => v.toLowerCase().indexOf(val.toLowerCase()) === 0) : [];
  }

}
