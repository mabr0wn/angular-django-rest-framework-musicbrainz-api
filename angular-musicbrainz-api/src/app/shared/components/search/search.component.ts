// Angular
import {
  Component,
  OnInit,
  OnDestroy,
  Output,
  EventEmitter,
  ChangeDetectionStrategy
 } from '@angular/core';
import { FormControl } from '@angular/forms';
// RxJs
import { 
  Observable, 
  Subject
} from 'rxjs';
import {
  map,
  startWith, 
  takeUntil 
} from 'rxjs/operators';
// Dummy data
import { SAMPLE_RESULTS } from '../../../sample-results';
// Model
import { Album } from '@core/models/album';
// Local
import { SearchParams} from '@shared/search-params';
import { SearchPresenter } from '@shared/presenters/search.presenter';

@Component({
  selector: 'app-search-component',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [SearchPresenter],
})
export class SearchComponent implements OnInit, OnDestroy {
  @Output() search: EventEmitter<SearchParams> = new EventEmitter<SearchParams>();
  private ngUnsubscribe: Subject<any> = new Subject();
  //
  queryString: string;
  searching: boolean;
  searchType: string;
  searchTerms: Subject<SearchParams> = new Subject<SearchParams>();
  //
  searchControl: FormControl;
  filteredResults$: Observable<string[]>;
  results = SAMPLE_RESULTS;


  constructor(private presenter: SearchPresenter) { }

  // ngOnInit(): void {
  //   this.searchControl = new FormControl('');
  //   this.filteredResults$ = this.searchControl.valueChanges.pipe(
  //     startWith(''),
  //     map(val => this.filterResults(val)),
  //     map(val => val.slice(0, 4)));
  // }

  ngOnInit(): void {
    this.presenter.searchTerms$.pipe(
      // complete when component is destroyed
      takeUntil(this.searchTerms),
    ).subscribe(() => this.search.emit());
  }

  searchFor(): void {
    this.queryString = this.queryString ? this.queryString.trim() : null;
    if (this.queryString) {
      this.searching = true;
      this.presenter.search();
      this.queryString = null;
    }
  }

  filterResults(val: string): string[] {
    return val ? this.results.filter(v => v.toLowerCase().indexOf(val.toLowerCase()) === 0) : [];
  }

  ngOnDestroy() {
    this.ngUnsubscribe.next();
    this.ngUnsubscribe.complete();
    this.searchTerms.next();
    this.searchTerms.complete();
  }

}
