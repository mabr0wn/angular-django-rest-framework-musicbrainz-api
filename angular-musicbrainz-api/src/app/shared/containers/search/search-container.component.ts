/* tslint:enable */
import {
  Component,
  OnInit,
  ChangeDetectionStrategy
} from '@angular/core';
// RxJS
import {
  of,
  Observable,
  Subject
 } from 'rxjs';
 import { 
   map,
   switchMap,
   distinctUntilChanged 
  } from 'rxjs/operators';
// Services
import { SearchService } from '@core/services/search/search.service';
// Local
import { SearchParams } from './search-params';
import { Album } from '@core/models/album';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'app-search-container',
  template: `
    <app-search-component (search)="search()">
      [results]="results | async"
    </app-search-component>
  `
})
/* tslint:disable */
export class SearchContainer implements OnInit {
  /* tslint:enable */
  albums: Observable<Album[]>;

  // search vars
  queryString: string;
  searching: boolean;
  searchType: string;


  searchTerms: Subject<SearchParams> = new Subject<SearchParams>();
  results$: Observable<Album[]>;

  private ngUnsubscribe: Subject<any> = new Subject();

  constructor(private searchService: SearchService) { }

  ngOnInit(): void {
    this.searching = false;
    this.searchType = 'release';
    this.searchTerms.pipe(
      distinctUntilChanged((params1, params2) => params2.equals(params1)),
      switchMap(
        (params) => this.searchService.searchAlbums(params.term, params.type),
        (_params, albums, _outerIdx, _innerIdx) => of(albums)
       )
    ).subscribe((albums) => {
      console.log(albums);
      this.results$ = albums;
      this.searching = false;
    });
  }

  // Push a search term into the observable stream.
  search(): void {
    this.searchTerms.next(new SearchParams(this.queryString, this.searchType));
  }

  filterResults(title: string) {
    return title ? this.results$.pipe(
      map(val => val.filter(v => v.title.indexOf(title) === 0))
    ) : [];
  }
}
