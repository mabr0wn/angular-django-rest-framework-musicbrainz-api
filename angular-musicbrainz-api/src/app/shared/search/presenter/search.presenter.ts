// RxJs
import { Subject } from 'rxjs';
// Local
import { SearchParams } from '@shared/search/search-params';

export class SearchPresenter {
  searchTerms: Subject<SearchParams> = new Subject();



  search(term): void {
    // Log the term for debugging purposes
    console.log(term, '@presenter');
    this.searchTerms.next(term);
  }
}
