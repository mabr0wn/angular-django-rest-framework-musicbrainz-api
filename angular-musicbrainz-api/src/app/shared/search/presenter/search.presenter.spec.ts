import { fakeAsync,  tick } from '@angular/core/testing';
import { Subscription } from 'rxjs';


import { SearchPresenter } from './search.presenter';

describe(SearchPresenter.name, () => {
    let presenter: SearchPresenter;


    beforeEach(() => {
        presenter = new SearchPresenter();
    });

    describe('emit search terms', () => {
        let searchTermsSubscription: Subscription;
        const searchTerm = {
            searchTerm: jest.fn((term) => `${term}`)
        };

        beforeEach(() => {
          searchTermsSubscription = presenter.searchTerms
            .subscribe(searchTerm.searchTerm);
        });

        afterEach(() => {
            searchTermsSubscription.unsubscribe();
        });

        test('it will descibe an album search', fakeAsync(() => {
            const tupac = 'tupac';

            presenter.search(tupac);

            expect(searchTerm.searchTerm).toHaveBeenCalledTimes(1);
            expect(searchTerm.searchTerm).toHaveBeenCalledWith(tupac);
        }));
    });
});
