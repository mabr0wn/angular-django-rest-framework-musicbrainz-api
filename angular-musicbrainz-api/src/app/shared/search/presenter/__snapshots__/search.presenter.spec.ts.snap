// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPresenter emit search terms it will descibe an album search: tupac 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "tupac",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "tupac",
    },
  ],
}
`;
